drivers/temperature/adt75/adt75.c

```diff
@@ -0,0 +1,179 @@
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include "adt75.h"
+#include "no_os_delay.h"
+#include "no_os_error.h"
+#include "no_os_util.h"
+#include "no_os_units.h"
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+/******************************************************************************/
+
+/**
+ * @brief Read the raw value of a register.
+ * @param desc - The device structure.
+ * @param addr - The register's address.
+ * @param val - The value to be read.
+ * @return 0 in case of success, -EINVAL otherwise
+ */
+int adt75_reg_read(struct adt75_desc *desc, uint32_t addr, uint16_t *val)
+{
+	int reg_size;
+	int ret;
+
+	reg_size = (addr == ADT75_CONF_REG)? 1 : 2;
+
+	desc->buff[0] = no_os_field_get(ADT75_DATA_REG_MASK, addr);
+	ret = no_os_i2c_write(desc->comm_desc, desc->buff, 1, 0);
+	if (ret)
+		return ret;
+
+	ret = no_os_i2c_read(desc->comm_desc, desc->buff, reg_size, 1);
+	if (ret)
+		return ret;
+
+	if (reg_size == 2)
+		*val = no_os_get_unaligned_be16(desc->buff);
+	else
+		*val = desc->buff[0];
+
+	return 0;
+}
+
+/**
+ * @brief Write a raw value to a register.
+ * @param desc - The device structure.
+ * @param addr - The register's address.
+ * @param val - The value to be written.
+ * @return 0 in case of success, -EINVAL otherwise
+ */
+int adt75_reg_write(struct adt75_desc *desc, uint32_t addr, uint16_t val)
+{
+	int reg_size;
+
+	reg_size = (addr == ADT75_CONF_REG)? 1 : 2;
+
+	desc->buff[0] = no_os_field_get(ADT75_DATA_REG_MASK, addr);
+	no_os_put_unaligned_be16(val, &desc->buff[1]);
+
+	return no_os_i2c_write(desc->comm_desc, desc->buff,
+			       reg_size + 1, 1);
+}
+
+/**
+ * @brief Read the current temperature sample (in milliCelsius)
+ * @param desc - The device structure.
+ * @param val - The temperature value.
+ * @return 0 in case of success, -EINVAL otherwise
+ */
+int adt75_get_single_temp(struct adt75_desc *desc, int32_t *val)
+{
+	uint16_t reg_val;
+	int ret;
+
+	ret = adt75_reg_read(desc, ADT75_TEMP_VALUE_REG, &reg_val);
+	if (ret)
+		return ret;
+
+	/** Wait for a new conversion to finish */
+	no_os_mdelay(ADT75_CONV_DELAY_MS);
+
+	reg_val = no_os_field_get(ADT75_TEMP_MASK, reg_val);
+	*val = no_os_sign_extend32(reg_val, ADT75_SIGN_BIT);
+	*val *= MILLIDEGREE_PER_DEGREE;
```

Comment:
nit: can reduce here the number of lines: `*val = *val * MILLIDEGREE_PER_DEGREE / ADT75_TEMP_DIV;`